### PR TITLE
BUG: fix edge counting in shortest_path

### LINKED ISSUE #5144: RuntimeWarning on csgraph.shortest_path when edge lengths are NaN
The documentation for scipy.sparse.csgraph says that NaN entries will be treated as non-edges, the same as 0 or inf. But when using csgraph.shortest_path on a graph where some paths are impossible because of NaN edges, irrelevant warning messages are printed. These warning messages don't happen in the other cases.

```
>>> import numpy
>>> import scipy.sparse.csgraph as csgraph
>>> print csgraph.shortest_path(numpy.array([[0,numpy.inf],[numpy.inf,0]]))
[[  0.  inf]
 [ inf   0.]]
>>> print csgraph.shortest_path(numpy.array([[0,0],[0,0]]))
[[  0.  inf]
 [ inf   0.]]
>>> print csgraph.shortest_path(numpy.array([[0, numpy.nan], [numpy.nan, 0]]))
/Users/creidieki/.virtualenvs/bugtest/lib/python2.7/site-packages/numpy/ma/core.py:2200: RuntimeWarning: invalid value encountered in less_equal
  condition = umath.less_equal(mabs(xnew - value), atol + rtol * mabs(value))
bug.py:4: RuntimeWarning: invalid value encountered in greater
  print csgraph.shortest_path(numpy.array([[0,numpy.nan],[numpy.nan,0]]))
bug.py:4: RuntimeWarning: invalid value encountered in less
  print csgraph.shortest_path(numpy.array([[0,numpy.nan],[numpy.nan,0]]))
[[  0.  inf]
 [ inf   0.]]
```

Using numpy v1.9.2, scipy v0.16.0, python 2.7.10 in a clean virtualenv on os x 10.10.4.

### PR DESCRIPTION
**Pull request by krzysztofpioro**
#### What does this implement/fix?
In the shortest paths function(with method == 'auto') there was a bug in counting edges for dense graph representation. (non-existent edges marked with np.inf were counted, while existing edges with weight 0 were skipped). This could lead to the selection of a suboptimal algorithm.
I fixed it by splitting it into two cases:
* For dense masked representation I counted number of non-masked elements.
* For dense array representation I changed elements equal to inf and NaN to 0 and then I counted number of elements bigger than 0.

I also got rid of the warnings that popped up when non-existent edges were marked by NaN. Thus this fixes #5144.

### CODE REVIEW

**Review by tylerjereddy — COMMENTED** [review]
Is it worth adding a unit test for the "bug fix?" I guess that's a little tricky for an algorithm selection fix.

I suppose you could check that warnings are not raised anymore.

**Comment by krzysztofpioro** [comment]
I've added a test that checks that warnings are not raised anymore. As for unit test for "bug fix" I don't know how to write it.

**Comment by rgommers** [comment]
@krzysztofpioro the test is failing:
```
    def test_NaN_warnings():
        with pytest.warns(None) as record:
            shortest_path(np.array([[0, 1], [np.nan, 0]]))
>       assert not record
E       assert not WarningsChecker(record=True)

record     = WarningsChecker(record=True)
```

**Comment by krzysztofpioro** [comment]
Fail was due to some PendingDeprecationWarning. I changed the test so that it now only check if RuntimeWarning is raised. The checks still don't pass, but it seems to me that this is not related to my changes.

File: scipy/sparse/csgraph/_shortest_path.pyx
Comment (pv) [inline scipy/sparse/csgraph/_shortest_path.pyx]: Modifies input --- you probably can write instead
```
edges = csgraph[np.isfinite(csgraph)]
```
Should you also then write `Nk = edges.size` --- `np.sum(csgraph > 0)` counts also inf entries?

Comment (krzysztofpioro) [reply scipy/sparse/csgraph/_shortest_path.pyx]: While I was trying to make the changes you suggested, I noticed that I forgot to count the negative edges(I counted positive values ​​instead of non-zero ones). I fixed it now and removed the input modification.

File: scipy/sparse/csgraph/_shortest_path.pyx
Comment (perimosocordiae) [inline scipy/sparse/csgraph/_shortest_path.pyx]: This will fail for certain sparse matrix formats. CSR, CSC, and COO should work, but DIA, LIL, BSR, and DOK are likely to cause trouble. I'd add a test first to confirm the bug, then add a conditional `csgraph.tocoo()` conversion before extracting edges.

File: scipy/sparse/csgraph/_shortest_path.pyx
Comment (perimosocordiae) [inline scipy/sparse/csgraph/_shortest_path.pyx]: the outer parens aren't needed here.

File: scipy/sparse/csgraph/tests/test_shortest_path.py
Comment (perimosocordiae) [inline scipy/sparse/csgraph/tests/test_shortest_path.py]:
```
for r in record:
  assert r.category ...
```

**Review by perimosocordiae — CHANGES_REQUESTED** [review]

File: scipy/sparse/csgraph/_shortest_path.pyx
Comment (krzysztofpioro) [reply scipy/sparse/csgraph/_shortest_path.pyx]: While I was writing a test for this bug I found that `floyd_warshall` throws ValueError when graph is passed as sparse matrix other than LIL, CSR or CSC.

Here is the code that produces this error:

```
G_dense = np.array([[0, 2, 1],
                    [0, -1, 0],
                    [0, 0, 0]])
G_sparse = coo_matrix(G_dense)
shortest_path(G_sparse)
```
Traceback:
```
  File "_shortest_path.pyx", line 189, in scipy.sparse.csgraph._shortest_path.shortest_path
  File "_shortest_path.pyx", line 298, in scipy.sparse.csgraph._shortest_path.floyd_warshall
  File "/home/piurko/TCS/PZ/scipy/scipy/sparse/csgraph/_validation.py", line 29, in validate_graph
    csgraph = csgraph_to_dense(csgraph, null_value=null_value_out)
  File "_tools.pyx", line 320, in scipy.sparse.csgraph._tools.csgraph_to_dense
ValueError: csgraph must be lil, csr, or csc format
```
For this reason I'm not sure what to do with the test you asked for. Should I just run `shortest_path` function with some graph (for which other algorithm than FW will be selected) in every sparse matrix format?

Comment (perimosocordiae) [reply scipy/sparse/csgraph/_shortest_path.pyx]: Oh, I didn't realize we had that restriction in place. Seems a bit too strict, but that's an issue for another time.

It should be sufficient to test with a LIL-format matrix.

File: scipy/sparse/csgraph/tests/test_shortest_path.py
Comment (perimosocordiae) [inline scipy/sparse/csgraph/tests/test_shortest_path.py]: Ideally we'd check that the result is correct here as well.

File: scipy/sparse/csgraph/tests/test_shortest_path.py
Comment (perimosocordiae) [inline scipy/sparse/csgraph/tests/test_shortest_path.py]: For completeness, could you loop over CSR, CSC, COO, and LIL format here?

File: scipy/sparse/csgraph/_shortest_path.pyx
Comment (perimosocordiae) [inline scipy/sparse/csgraph/_shortest_path.pyx]: Since we already know it's a sparse matrix, we can do:

```python
if csgraph.format in ('csr', 'csc', 'coo'):
    edges = csgraph.data
else:
    edges = csgraph.tocoo().data
```

(I suggest using `tocoo()` because the data array is equivalent and converting to COO is often faster.)

**Review by perimosocordiae — CHANGES_REQUESTED** [review]

File: scipy/sparse/csgraph/tests/test_shortest_path.py
Comment (krzysztofpioro) [reply scipy/sparse/csgraph/tests/test_shortest_path.py]: Are you sure about COO? Restrictions in `csgraph_to_dense` alows only CSR, CSC and LIL. However, the test would still pass with COO because Johnson is chosen for this graph.

And is there some cleaner way to iterate through this formats than copy paste?

**Comment by tylerjereddy** [comment]
@krzysztofpioro @perimosocordiae  What do we think about the status of this PR with ~1 week left until the next release branch?

**Comment by krzysztofpioro** [comment]
Oh, I was waiting for a @perimosocordiae reply about COO and I completely forgot about this PR. Sorry for that. 

I remember I had two commits ready, so I'll push them in a moment. However, I don't have time to work on this, so if more changes are required I will probably close this PR.

File: scipy/sparse/csgraph/tests/test_shortest_path.py
Comment (krzysztofpioro) [reply scipy/sparse/csgraph/tests/test_shortest_path.py:333]: I've added only CSR and CSC tests.

**Comment by perimosocordiae** [comment]
Thanks, I'll take another look soon. If there are more changes to be made, I can apply them on top of the existing commits.

**Review by perimosocordiae — APPROVED** [review]

**Review by larsoner — APPROVED** [review]
Azure failures appear to be the unrelated C/D drive issue

**Comment by tylerjereddy** [comment]
2 core devs +1 to merge here and unrelated CI fail is fixed in `master`--let's put this in

**Comment by tylerjereddy** [comment]
Thanks @krzysztofpioro  & reviewers

**Pull request merged**
